### PR TITLE
Add volatile casts and use them to print case univ and parameter data

### PR DESCRIPTION
**Pull request by SkySkimmer**
Overlays:
- https://github.com/QuickChick/QuickChick/pull/319

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Probably should have some flag to control the case printing

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: doc/sphinx/language/extensions/match.rst [hunk 1]
```diff
@@ -244,6 +244,33 @@ Printing of the elimination predicate
...
+   variables bound by `in` and `as`, `match` terms contain the
+   polymorphic universe instance and the parameters of the inductive
+   which is being matched. When this flag is on (it is off by
+   default), this information is displayed as a volatile cast around
+   the match discriminee.
+
+   When the match relies on :flag:`Definitional UIP` the indices are also subterms of the `match` term and are displayed when this flag is on.
+   Otherwise they are not subterms and are displaye as holes `_` when this flag is on.
+
+   .. coqtop:: all
```
Comment (jfehrle) [inline doc/sphinx/language/extensions/match.rst:262]:
```suggestion
   .. example:
   .. coqtop:: in
```

Then indent all of the example lines by another 3 spaces.  Use "in" for the first couple lines to make the example shorter.

File: doc/sphinx/language/extensions/match.rst [hunk 2]
```diff
@@ -244,6 +244,33 @@ Printing of the elimination predicate
...
+
+      Set Printing Match All Subterms.
+      Set Printing Universes.
+
+      Polymorphic Inductive eqT@{u} {A:Type@{u}} (a:A) : A -> Type@{u} := reflT : eqT a a.
+      Print eqT_rect.
+
+      Set Definitional UIP.
+      Inductive seq {A} (a:A) : A -> SProp := srefl : seq a a.
+      Print seq_rect.
```
Comment (jfehrle) [inline doc/sphinx/language/extensions/match.rst:272]:
```suggestion
   .. coqtop:: all
   
      Print eqT_rect.

   .. coqtop:: in
   
      Set Definitional UIP.
      Inductive seq {A} (a:A) : A -> SProp := srefl : seq a a.

   .. coqtop:: all
      
      Print seq_rect.
```
Omit uninteresting output to make the example shorter.

I expected to see a volatile cast in the Print, i.e. `@term10 :> @type`.  Or am I misunderstanding?

The example would be a lot better if included both Prints with and without `Printing Match All Subterms` so readers can see the differences.

File: doc/sphinx/language/core/definitions.rst [hunk 3]
```diff
@@ -52,6 +53,9 @@ to type check that :n:`@term10` has type :n:`@type` (see :tacn:`vm_compute`).
 :n:`@term10 <<: @type` specifies that compilation to OCaml will be used
 to type check that :n:`@term10` has type :n:`@type` (see :tacn:`native_compute`).
 
+:n:`@term10 :> @type` enforces the type of :n:`@term10` to be
+:n:`@type` without leaving a trace in the produced value.
```
Comment (jfehrle) [inline doc/sphinx/language/core/definitions.rst:57]:
```suggestion
:n:`@term10 :> @type` enforces the type of :n:`@term10` to be
:n:`@type` without leaving a trace in the produced value.
This is a :gdef:`volatile cast`.
```

File: doc/sphinx/language/extensions/match.rst [hunk 4]
```diff
@@ -244,6 +244,33 @@ Printing of the elimination predicate
...
+Printing of hidden subterms
++++++++++++++++++++++++++++
+
+.. flag:: Printing Match All Subterms
+
+   In order to be able to cheaply reconstruct the types of the
+   variables bound by `in` and `as`, `match` terms contain the
+   polymorphic universe instance and the parameters of the inductive
+   which is being matched. When this flag is on (it is off by
+   default), this information is displayed as a volatile cast around
```
Comment (jfehrle) [inline doc/sphinx/language/extensions/match.rst:256]:
```suggestion
   default), this information is displayed as a :term:`volatile cast` around
```

**Review by jfehrle — COMMENTED** [review]
Wording suggestions

**Comment by JasonGross** [comment]
How does this interact with notation printing?  That is, can I make a printing notation like:
```
Notation "x $ T" := (x :> T).
```
?

**Comment by SkySkimmer** [comment]
>How does this interact with notation printing?

You can declare the notation, it will be used when printing non-elaborated terms (eg `let c := uconstr:(True :> True) in idtac c.`) and when using the flag introduced in this PR.

File: doc/sphinx/language/extensions/match.rst (hunk 2 above)
Comment (SkySkimmer) [reply doc/sphinx/language/extensions/match.rst]: >I expected to see a volatile cast in the Print, i.e. @term10 :> @type. Or am I misunderstanding?

Screenshot_20230120_204140

Comment (jfehrle) [reply doc/sphinx/language/extensions/match.rst]: 👍

**Review by herbelin — COMMENTED** [review]
Very nice.

**Comment by SkySkimmer** [comment]
@herbelin would o=you like to assign/merge?

**Comment by herbelin** [comment]
OK, but ping me on Monday if no news by then (and no problem is someone has time before).

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Comment by herbelin** [comment]
Thanks!

**Pull request merged**